catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java

```diff
@@ -0,0 +1,129 @@
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+
+package org.codice.alliance.catalog.plugin.security.audit;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.collections.MapUtils;
+
+import ddf.catalog.data.AttributeDescriptor;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.operation.CreateRequest;
+import ddf.catalog.operation.DeleteRequest;
+import ddf.catalog.operation.DeleteResponse;
+import ddf.catalog.operation.QueryRequest;
+import ddf.catalog.operation.QueryResponse;
+import ddf.catalog.operation.ResourceRequest;
+import ddf.catalog.operation.ResourceResponse;
+import ddf.catalog.operation.UpdateRequest;
+import ddf.catalog.plugin.AccessPlugin;
+import ddf.catalog.plugin.StopProcessingException;
+import ddf.catalog.util.impl.Requests;
+import ddf.security.common.audit.SecurityLogger;
+
+public class SecurityAuditPlugin implements AccessPlugin {
+
+    private List<String> auditAttributes = new ArrayList<>();
+
+    @Override
+    public CreateRequest processPreCreate(CreateRequest createRequest)
+            throws StopProcessingException {
+        return createRequest;
+    }
+
+    @Override
+    public UpdateRequest processPreUpdate(UpdateRequest updateRequest,
+            Map<String, Metacard> existingMetacards) throws StopProcessingException {
+        if (updateRequest != null && Requests.isLocal(updateRequest)) {
+            for (Map.Entry<Serializable, Metacard> entry : updateRequest.getUpdates()) {
+                Metacard updateMetacard = entry.getValue();
+                if (MapUtils.isNotEmpty(existingMetacards)) {
+                    Metacard existingMetacard = existingMetacards.get(entry.getKey()
+                            .toString());
+                    Set<AttributeDescriptor> attributeDescriptors = updateMetacard.getMetacardType()
+                            .getAttributeDescriptors();
+                    for (AttributeDescriptor descriptor : attributeDescriptors) {
+                        String descriptorName = descriptor.getName();
+                        if (auditAttributes.contains(descriptorName)) {
+                            if (updateMetacard.getAttribute(descriptorName) != null) {
+                                Serializable newValue = updateMetacard.getAttribute(descriptorName)
+                                        .getValue();
+                                if (existingMetacard.getAttribute(descriptorName) != null) {
```

Comment:
Since getAttribute is a fast map lookup we could just get both the existing and update attribute and then check to see if they are null in the same if statement as the `auditAttributes.contains(descriptorName)` and then do a `continue` if the descriptorName is not an audit attribute or one of the metacards doesn't have the attribute.